File: 938_range_sum_bst.cpp
/*
[mst] 938_range_sum_bst.cpp
Given the root node of a binary search tree and two integers low and high, return the sum of values of all nodes with a value in the inclusive range [low, high].

Gains:
-Binary trees behaviour
-Recursive aproach cpp

features, changelog:
-2022.12: -initial draft
		  -recursive solution submission
*/ 


////////////////// LIBS
#include <iostream>            // usage of console prints
using namespace std;


////////////////// DECL_IMPL

class TreeNode {
private:
	
public:
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(-1), left(nullptr), right(nullptr){}	// here invalid value is -1
	TreeNode(int x) : val(x), left(nullptr), right(nullptr){}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right){}

	void printInOrder(){
		if (left) { left->printInOrder(); }
		cout << val << endl;
		if (right) { right->printInOrder(); }
	}

	// binary search tree insertion
	void insert(int x) {
		// empty tree case
		if (val == -1) {
			val = x;
			return;
		}
		
		if (x <= val) {	// value lesser than the root will be routed left
			if (left == nullptr){ left = new TreeNode(x); }	// empty left node
			else {
				left->insert(x);	// non empty left node
			}
		}
		else {	// value higherr than the root will be routed right
			if (right == nullptr){ right = new TreeNode(x); }	// empty right node
			else {
				right->insert(x);	// non empty right node
			}			
		}
	}

	// insertion overloaded for array-to-tree conversion
	void insert(int arr[], int size) {
		if (size == 0) return;
		for(int i = 0; i<size; i++) {
			this->insert(arr[i]);
		}
	}
};

// traverse the tree in-order and add the node value if within range
//
// sub: 82%T 60%S
class Solution {
public:
	int rangeSumBST(TreeNode* root, int low, int high) {
        if (root == nullptr){
			return 0;
		}

		int to_add = 0;
		if ((root->val <= high) && (root->val >= low)) to_add = root->val;

		return rangeSumBST(root->left, low, high) + rangeSumBST(root->right, low, high) + to_add;
    }

};


////////////////// DRIVER
int main()
{
	TreeNode tree1;

	int array[6] = {10,5,15,3,7,18};
	tree1.insert(array,6);
	tree1.printInOrder();

	Solution sol;
	cout << "result: " << sol.rangeSumBST(&tree1, 7, 15);

	cin.get(); // pseudo-pause the console
	return 0;
}

File: done/938_range_sum_bst.py
##!/usr/bin/env python
'''
    @author [mst]
    @brief  leetcode problems series
    938_range_sum_bst.py
    Given the root node of a binary search tree and two integers low and high,
    return the sum of values of all nodes with a value in the inclusive range [low, high].
    
    gains:
    -python tree structure
    -basic tree traversal

    features, changelog:
    -2023.01: -submission

    @version 0.1 2022.11
'''


            
################## DECL_IMPL

# Definition for a binary tree node. (taken from leetcode problems)
class TreeNode:
    def __init__(self, val=-1, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def print_inorder(self):
        if self.left:
            self.left.print_inorder()
        print(f"{self.val} ", end='')
        if self.right:
            self.right.print_inorder()
        return

    # insert single node
    def insert(self, x):
        # empty node insertion
        if self.val == -1:
            self.val = x
            return

        # bst insertions
        if x <= self.val:
            if self.left:
                self.left.insert(x)
            else:
                self.left = TreeNode(x)
        else:
            if self.right:
                self.right.insert(x)
            else:
                self.right = TreeNode(x)
        return

    # array to bst
    def insert_arr(self, ar : []):
        if len(ar) == 0: return

        for n in ar:
            self.insert(n)


# traverse the tree and add the node value if within range
#
# sub: 73%T 50%S
class Solution:
    def rangeSumBST(self, root: [TreeNode], low: int, high: int) -> int:
        if root == None:
            return 0

        adding = 0
        if (root.val <= high) and (root.val >= low):
            adding = root.val

        return self.rangeSumBST(root.left, low, high) + self.rangeSumBST(root.right, low, high) + adding


################## DRIVER
def main():
    sol = Solution()

    # root = [10,5,15,3,7,18]
    # low = 7
    # high = 15

    root = [10,5,15,3,7,13,18,1,6]
    low = 6
    high = 10

    # testing
    head = TreeNode()
    head.insert_arr(root)
    res = sol.rangeSumBST(head, low, high)
    print(f"{root=} {low=}{high=}, res= {res=}")


    
    


if __name__ == ("__main__"):
    main()
